core/src/test/scala/quasar/destination/gbq/GBQDestinationSpec.scala

```diff
@@ -0,0 +1,265 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.destination.gbq
+
+import slamdata.Predef._
+
+import quasar.api.destination.DestinationError.InitializationError
+import quasar.api.destination.{Destination, ResultSink}
+import quasar.api.table.{ColumnType, TableColumn}
+import quasar.api.resource.{ResourceName, ResourcePath}
+import quasar.api.destination.{Destination, DestinationError, DestinationType}
+import quasar.connector.ResourceError
+import quasar.contrib.proxy.Search
+import quasar.contrib.scalaz.MonadError_
+
+import quasar.EffectfulQSpec
+
+import argonaut._, Argonaut._
+
+import cats.effect.{ConcurrentEffect, IO, Resource, Timer}
+import cats.implicits._
+
+import eu.timepit.refined.auto._
+
+import fs2.{Stream, text}
+
+import java.nio.file.{Files, Paths}
+import java.nio.charset.StandardCharsets.UTF_8
+
+import org.http4s.argonaut.jsonEncoderOf
+import org.http4s.client.asynchttpclient.AsyncHttpClient
+import org.http4s.client.Client
+import org.http4s.{
+  AuthScheme, 
+  Credentials, 
+  EntityEncoder, 
+  MediaType,
+  Method, 
+  Request, 
+  Status,
+  Uri}
+import org.http4s.headers.Authorization
+import org.http4s.headers.`Content-Type`
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import scala.util.Right
+import scalaz.{-\/,\/-}
+
+import shims._
+
+object GBQDestinationSpec extends EffectfulQSpec[IO] {
+  sequential
+
+  implicit val timer: Timer[IO] = IO.timer(ExecutionContext.global)
+
+  val testProject = "travis-ci-reform-test-proj"
+  val testDataset = "testdataset"
+  val authCfgPath = Paths.get(getClass.getClassLoader.getResource("gbqAuthFile.json").toURI)
+  val authCfgString: String = new String(Files.readAllBytes(authCfgPath), UTF_8)
+  val cfg = config(authCfg = Some(authCfgString), Some(testProject), Some(testDataset))
+  val resource = mkClient[IO]
+
+  "csv link" should {
+    "reject empty paths with NotAResource" >>* {
+      csv(cfg) { sink =>
+        val p = ResourcePath.root()
+        val r = sink.run(p, List(TableColumn("a", ColumnType.Boolean)), Stream.empty).compile.drain
+
+        MRE.attempt(r).map(_ must beLike {
+          case -\/(ResourceError.NotAResource(p2)) => p2 must_=== p
+        })
+      }
+    }
+
+    "successfully upload table" >>* {
+      val dst = ResourcePath.root() / ResourceName("foo") / ResourceName("bar.csv")
+      val data = Stream("col1,col2\r\nstuff,true\r\n").through(text.utf8Encode)
+
+      csv(cfg) { sink =>
+          val r = sink.run(dst, List(TableColumn("a", ColumnType.String), TableColumn("b", ColumnType.Boolean)), data).compile.drain
+          //TODO: give BigQuery to make it aware of newly pushed table, is there something else we can do?
+          java.lang.Thread.sleep(5000)
```

Comment:
Never ever `Thread.sleep`. Use `Timer.sleep` instead if you really need to wait in this fashion.